packages/circuits/test/proveSubsidyKey.test.ts

```diff
@@ -0,0 +1,320 @@
+    const epochKeyModed = epochKey % BigInt(2 ** _epochTreeDepth)
+    return epochKeyModed
+}
+
+const genCircuitInput = (
+    id: crypto.ZkIdentity,
+    epoch: number,
+    reputationRecords,
+    attesterId,
+    notEpochKey,
+    _minRep?
+) => {
+    const minRep = _minRep ?? 0
+    if (reputationRecords[attesterId] === undefined) {
+        reputationRecords[attesterId] = new Reputation(
+            BigInt(0),
+            BigInt(0),
+            BigInt(0),
+            BigInt(0)
+        )
+    }
+
+    // User state tree
+    const userStateTree = new crypto.SparseMerkleTree(
+        USER_STATE_TREE_DEPTH,
+        crypto.hash5([BigInt(0), BigInt(0), BigInt(0), BigInt(0), BigInt(0)])
+    )
+    for (const attester of Object.keys(reputationRecords)) {
+        userStateTree.update(
+            BigInt(attester),
+            reputationRecords[attester].hash()
+        )
+    }
+    const userStateRoot = userStateTree.root
+    const USTPathElements = userStateTree.createProof(BigInt(attesterId))
+
+    // Global state tree
+    const GSTree = new crypto.IncrementalMerkleTree(GLOBAL_STATE_TREE_DEPTH)
+    const commitment = id.genIdentityCommitment()
+    const hashedLeaf = crypto.hashLeftRight(commitment, userStateRoot)
+    GSTree.insert(hashedLeaf)
+    const GSTreeProof = GSTree.createProof(0) // if there is only one GST leaf, the index is 0
+    const GSTreeRoot = GSTree.root
+
+    const circuitInputs = {
+        epoch,
+        identity_nullifier: id.identityNullifier,
+        identity_trapdoor: id.trapdoor,
+        user_tree_root: userStateRoot,
+        GST_path_index: GSTreeProof.pathIndices,
+        GST_path_elements: GSTreeProof.siblings,
+        attester_id: attesterId,
+        pos_rep: reputationRecords[attesterId].posRep,
+        neg_rep: reputationRecords[attesterId].negRep,
+        graffiti: reputationRecords[attesterId].graffiti,
+        sign_up: reputationRecords[attesterId].signUp,
+        UST_path_elements: USTPathElements,
+        minRep,
+        notEpochKey,
+    }
+    return crypto.stringifyBigInts(circuitInputs)
+}
+
+const genProofAndPublicSignals = async (inputs: any): Promise<any> => {
+    const circuitWasmPath = path.join(
+        __dirname,
+        `../zksnarkBuild/${circuitName}.wasm`
+    )
+    const zkeyPath = path.join(__dirname, `../zksnarkBuild/${circuitName}.zkey`)
+    const { proof, publicSignals } = await snarkjs.groth16.fullProve(
+        inputs,
+        circuitWasmPath,
+        zkeyPath
+    )
+    return { proof, publicSignals }
+}
+
+const verifyProof = async (publicSignals, proof): Promise<boolean> => {
+    const vkey = require(path.join(
+        __dirname,
+        `../zksnarkBuild/${circuitName}.vkey.json`
+    ))
+    return snarkjs.groth16.verify(vkey, publicSignals, proof)
+}
+
+const calcSubsidyKey = (id, epoch) => {
+    return crypto.hash2([
+        BigInt(id.identityNullifier) + BigInt(NUM_EPOCH_KEY_NONCE_PER_EPOCH),
+        BigInt(epoch),
+    ])
+}
```

Comment:
it should be able to use a `genEpochKey` to generate epoch key
e.g.
```typescript
const calcSubsidyKey = (id, epoch) => {
    const nonce = NUM_EPOCH_KEY_NONCE_PER_EPOCH
    return genEpochKey(id, epoch, nonce)
}
```

But the user will never receive reputation with this epoch key because the nonce cannot be equal to `NUM_EPOCH_KEY_NONCE_PER_EPOCH`